src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs

```diff
@@ -300,103 +300,40 @@ internal void Clear()
 
     internal static class SqlEnvChangePool
     {
-        private static int _count;
-        private static int _capacity;
-        private static SqlEnvChange[] _items;
-#if DEBUG
-        private static string[] _stacks;
-#endif 
-
-        static SqlEnvChangePool()
+        private protected struct SqlEnvChangeWrapper
         {
-            _capacity = 10;
-            _items = new SqlEnvChange[_capacity];
-            _count = 1;
-#if DEBUG
-            _stacks = new string[_capacity];
-#endif 
+            public SqlEnvChange Item;
         }
 
+        private static readonly SqlEnvChangeWrapper[] _items = new SqlEnvChangeWrapper[Environment.ProcessorCount * 2];
+
         internal static SqlEnvChange Allocate()
         {
-            SqlEnvChange retval = null;
-            lock (_items)
+            SqlEnvChangeWrapper[] items = _items;
+            for (int index = 0; index < items.Length; index++)
             {
-                while (_count > 0 && retval is null)
+                SqlEnvChange item = items[index].Item;
+                if (item != null && Interlocked.CompareExchange(ref items[index].Item, null, item) == item)
                 {
-                    int count = _count; // copy the count we think we have
-                    int newCount = count - 1; // work out the new value we want
-                                              // exchange _count for newCount only if _count is the same as our cached copy
-                    if (count == Interlocked.CompareExchange(ref _count, newCount, count))
-                    {
-                        // count is now the previous value, we're the only thread that has it, so count-1 is safe to access
-                        Interlocked.Exchange(ref retval, _items[count - 1]);
-                    }
-                    else
-                    {
-                        // otherwise the count wasn't what we expected, spin the while and try again.
-                    }
+                    return item;
                 }
             }
-            if (retval is null)
-            {
-                retval = new SqlEnvChange();
-            }
-            return retval;
+            return new SqlEnvChange();
         }
 
-        internal static void Release(SqlEnvChange item, [Runtime.CompilerServices.CallerMemberName] string caller = null)
+        internal static void Release(SqlEnvChange item)
         {
             if (item is null)
             {
-                Debug.Fail("attenpting to release null packet");
+                Debug.Fail("attenpting to release null item");
                 return;
             }
             item.Clear();
+            SqlEnvChangeWrapper[] items = _items;
+            for (int index = 0; index < items.Length && Interlocked.CompareExchange(ref items[index].Item, item, null) != null; index++)
```

Comment:
This is pretty long. My preference would be `if`/`break` to make the code faster to read, even though it's less clever.

Reply:
Reworked to a better format.